packages/@aws-cdk/aws-codepipeline/lib/private/stage.ts

```diff
@@ -105,6 +106,28 @@ export class Stage implements IStage {
     return rule;
   }
 
+  public notifyOn(id: string, options: notifications.NotifyOnEventOptions = {}): notifications.INotificationRule {
+    const rule = new notifications.NotificationRule(this.scope, id, {
+      ...options,
+      source: this.pipeline,
+    });
+    return rule;
+  }
+
+  public notifyOnStateChange(id: string, options: notifications.NotifyOptions = {}): notifications.INotificationRule {
```

Comment:
Can you explain something to me? I don't see any mention of a particular Stage here. Does this mean this event triggers for every stage transition in the pipeline?

Reply:
Thanks, this is a good question, and I found a big problem that [`CfnNotificationRule` is based on `pipeline` , not `stage` and `action` in pipeline case](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codestarnotifications-notificationrule.html#cfn-codestarnotifications-notificationrule-resource), which means you cannot set any event id in a particular Stage or a particular Action.

Currently, our design will make many many duplicate notification, if we design the API in `stage level` and `action level` . 

For example, if we set `codepipeline-pipeline-action-execution-succeeded` event id which is an `action level` event to one of all actions, then all of actions will trigger notification when action execution succeeded in `pipeline`; likewise, if we set `codepipeline-pipeline-stage-execution-succeeded` event id which is a `stage level` event to one of all stages, then all of stages will trigger notification when stage execution succeeded in `pipeline`.

So my proposal by the previous thoughts, you can consider something in `pipeline` like this:

```ts
notifyOnPipelineStateChange(id, target, options);
notifyOnActionStateChange(id, target, options);
notifyOnStageStateChange(id, target, options);
```

Our usage in reality that we just let `pipeline` be able to notify, and we known it is based on pipeline, not stages and actions.

Does it make sense?

Or do you have any ideas?

Reply:
Ok. Makes sense. I feel like this is a really weird design, but it's not like we can change how the underlying service works.

One suggestion from me: let's change the names slightly, so that it's obvious this will trigger for any Stage/Action:

```ts
notifyOnAnyStageStateChange(id, target, options);
notifyOnAnyActionStateChange(id, target, options);
```

Reply:
I know you want to be able to control every stages and every actions, but this is a limitation of `CfnNotificationRule`.

Maybe This is a aweird design, and I don't have no idea to make it better..., but a different perspective, this usage is very similar to `CodeBuild`, in fact, when I am a CDK user, I think these APIs are very very useful.🙂

BTW, should I add `notifyOnAnyApprovalStateChange` in IPipeline? or do you have a nice name for it?

Reply:
> BTW, should I add `notifyOnAnyApprovalStateChange` in IPipeline? or do you have a nice name for it?

Sure, let's add it to `IPipeline` too. I would call it `notifyOnAnyManualApprovalStateChange()`.

Reply:
Ok. I will adjust.